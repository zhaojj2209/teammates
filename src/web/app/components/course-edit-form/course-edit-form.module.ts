import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { NgbTooltipModule } from '@ng-bootstrap/ng-bootstrap';
import { AjaxLoadingModule } from '../ajax-loading/ajax-loading.module';
import { LoadingRetryModule } from '../loading-retry/loading-retry.module';
import { LoadingSpinnerModule } from '../loading-spinner/loading-spinner.module';
import { CourseEditFormComponent } from './course-edit-form.component';

@NgModule({
  declarations: [
    CourseEditFormComponent,
  ],
  imports: [
    CommonModule,
    FormsModule,
    AjaxLoadingModule,
    NgbTooltipModule,
    LoadingRetryModule,
    LoadingSpinnerModule,
  ],
  exports: [
    CourseEditFormComponent,
  ],
})
export class CourseEditFormModule { }
