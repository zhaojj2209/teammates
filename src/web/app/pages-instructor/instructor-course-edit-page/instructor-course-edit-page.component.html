<tm-loading-retry [shouldShowRetry]="hasCourseLoadingFailed" [message]="'Failed to load course'" (retryEvent)="loadCourseInfo()">
  <div *tmIsLoading="isCourseLoading" class="card bg-light top-padded text-center">
  <tm-course-edit-form
    [formMode]="CourseEditFormMode.EDIT"
    [(formModel)]="courseFormModel"
    [resetFormEvent]="resetCourseFormEvent"
    (deleteCourseEvent)="deleteCourse()"
    (updateCourseEvent)="onSaveCourse()"
  ></tm-course-edit-form>
  </div>
</tm-loading-retry>
<tm-loading-retry [shouldShowRetry]="hasInstructorsLoadingFailed" [message]="'Failed to load instructors of course'" (retryEvent)="loadCourseInstructors()">
  <div class="right top-padded">
    <a class="small" tmRouterLink="/web/instructor/help" [queryParams]="{questionId: CoursesSectionQuestions.COURSE_INSTRUCTOR_ACCESS, section: Sections.courses}" rel="noopener noreferrer" target="_blank">
      <i class="fa fa-info-circle" aria-hidden="true"></i>
      More about configuring access permissions
    </a>
  </div>
  <div *tmIsLoading="isInstructorsLoading">
    <ng-container *ngFor="let panelDetail of instructorDetailPanels; let i = index">
      <tm-instructor-edit-panel
          [editMode]="EditMode.EDIT" [(instructor)]="panelDetail.editPanel" [instructorIndex]="i"
          [currInstructorCoursePrivilege]="currInstructorCoursePrivilege"
          (viewRolePrivilegeModel)="viewRolePrivilegeModel($event)"
          (deleteInstructor)="deleteInstructor(i)"
          (saveInstructor)="saveInstructor(i)"
          (cancelEditing)="cancelEditingInstructor(i)"
          (sendRemindJoinEmail)="resendReminderEmail(i)"
          [allSections]="this.allSections" [allSessions]="this.allSessions">
      </tm-instructor-edit-panel>
      <br/>
    </ng-container>
  </div>
</tm-loading-retry>
<div class="card-body text-center" *ngIf="!isAddingNewInstructor">
  <button id="btn-add-instructor" class="btn btn-success mx-2 my-1" type="button" (click)="isAddingNewInstructor = true" [disabled]="isCopyingInstructor">
    Add New Instructor
  </button>
  <button id="btn-copy-instructor" class="btn btn-info mx-2 my-1" type="button" (click)="copyInstructors()" [disabled]="!this.currInstructorCoursePrivilege.canModifyInstructor || isCopyingInstructor">
    <tm-ajax-loading *ngIf="isCopyingInstructor"></tm-ajax-loading>Copy Instructors from Other Courses
  </button>
</div>
<div *ngIf="isAddingNewInstructor" @collapseAnim>
  <tm-instructor-edit-panel [editMode]="EditMode.ADD" [(instructor)]="newInstructorPanel" [instructorIndex]="instructorDetailPanels.length"
                            [isSavingNewInstructor]="isSavingNewInstructor"
                            [currInstructorCoursePrivilege]="currInstructorCoursePrivilege"
                            (viewRolePrivilegeModel)="viewRolePrivilegeModel($event)"
                            (saveInstructor)="addNewInstructor()" (cancelEditing)="isAddingNewInstructor = false;"
                            [allSections]="this.allSections" [allSessions]="this.allSessions"></tm-instructor-edit-panel>
</div>
