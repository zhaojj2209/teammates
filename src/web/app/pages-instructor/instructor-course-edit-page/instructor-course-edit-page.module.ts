import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { RouterModule, Routes } from '@angular/router';
import { NgbTooltipModule } from '@ng-bootstrap/ng-bootstrap';
import { AjaxLoadingModule } from '../../components/ajax-loading/ajax-loading.module';
import { CourseEditFormModule } from '../../components/course-edit-form/course-edit-form.module';
import { LoadingRetryModule } from '../../components/loading-retry/loading-retry.module';
import { LoadingSpinnerModule } from '../../components/loading-spinner/loading-spinner.module';
import { PanelChevronModule } from '../../components/panel-chevron/panel-chevron.module';
import { TeammatesCommonModule } from '../../components/teammates-common/teammates-common.module';
import { TeammatesRouterModule } from '../../components/teammates-router/teammates-router.module';
import {
  CopyInstructorsFromOtherCoursesModalComponent,
} from './copy-instructors-from-other-courses-modal/copy-instructors-from-other-courses-modal.component';
import {
  CustomPrivilegeSettingPanelComponent,
} from './custom-privilege-setting-panel/custom-privilege-setting-panel.component';
import { InstructorCourseEditPageComponent } from './instructor-course-edit-page.component';
import { InstructorEditPanelComponent } from './instructor-edit-panel/instructor-edit-panel.component';
import { ViewRolePrivilegesModalComponent } from './view-role-privileges-modal/view-role-privileges-modal.component';

const routes: Routes = [
  {
    path: '',
    component: InstructorCourseEditPageComponent,
  },
];

/**
 * Module for instructor course edit page component.
 */
@NgModule({
  declarations: [
    InstructorCourseEditPageComponent,
    InstructorEditPanelComponent,
    ViewRolePrivilegesModalComponent,
    CustomPrivilegeSettingPanelComponent,
    CopyInstructorsFromOtherCoursesModalComponent,
  ],
  exports: [
    InstructorCourseEditPageComponent,
  ],
  imports: [
    CommonModule,
    FormsModule,
    RouterModule.forChild(routes),
    NgbTooltipModule,
    TeammatesCommonModule,
    LoadingSpinnerModule,
    AjaxLoadingModule,
    LoadingRetryModule,
    TeammatesRouterModule,
    PanelChevronModule,
    CourseEditFormModule,
  ],
  entryComponents: [
    ViewRolePrivilegesModalComponent,
  ],
})
export class InstructorCourseEditPageModule { }
