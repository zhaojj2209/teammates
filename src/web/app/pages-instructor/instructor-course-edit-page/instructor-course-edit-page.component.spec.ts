import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { FormsModule } from '@angular/forms';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { RouterTestingModule } from '@angular/router/testing';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
import { of } from 'rxjs';
import { CourseService } from '../../../services/course.service';
import { InstructorService } from '../../../services/instructor.service';
import { Course, Instructor, InstructorPermissionRole, JoinState } from '../../../types/api-output';
import { InstructorCreateRequest } from '../../../types/api-request';
import { AjaxLoadingModule } from '../../components/ajax-loading/ajax-loading.module';
import { CourseEditFormComponent } from '../../components/course-edit-form/course-edit-form.component';
import { LoadingRetryModule } from '../../components/loading-retry/loading-retry.module';
import { LoadingSpinnerModule } from '../../components/loading-spinner/loading-spinner.module';
import { PanelChevronModule } from '../../components/panel-chevron/panel-chevron.module';
import { SimpleModalModule } from '../../components/simple-modal/simple-modal.module';
import { TeammatesCommonModule } from '../../components/teammates-common/teammates-common.module';
import { TeammatesRouterModule } from '../../components/teammates-router/teammates-router.module';
import {
  CopyInstructorsFromOtherCoursesModalComponent,
} from './copy-instructors-from-other-courses-modal/copy-instructors-from-other-courses-modal.component';
import {
  CustomPrivilegeSettingPanelComponent,
} from './custom-privilege-setting-panel/custom-privilege-setting-panel.component';
import { InstructorCourseEditPageComponent } from './instructor-course-edit-page.component';
import {
  InstructorEditPanel,
  InstructorEditPanelComponent,
} from './instructor-edit-panel/instructor-edit-panel.component';
import { ViewRolePrivilegesModalComponent } from './view-role-privileges-modal/view-role-privileges-modal.component';

const testCourse: Course = {
  courseId: 'exampleId',
  courseName: 'Example Course',
  institute: 'Test Institute',
  timeZone: 'UTC (UTC)',
  creationTimestamp: 0,
  deletionTimestamp: 1000,
};

const testInstructor1: Instructor = {
  courseId: 'exampleId',
  email: 'instructor1@gmail.com',
  joinState: JoinState.JOINED,
  name: 'Instructor 1',
};

const testInstructor2: Instructor = {
  courseId: 'exampleId',
  email: 'instructor2@gmail.com',
  joinState: JoinState.NOT_JOINED,
  name: 'Instructor 2',
};

const testInstructor3: Instructor = {
  courseId: 'exampleId',
  email: 'instructor3@gmail.com',
  joinState: JoinState.NOT_JOINED,
  name: 'Instructor 3',
};

const emptyInstructorPanel: InstructorEditPanel = {
  googleId: '',
  courseId: '',
  email: '',
  isDisplayedToStudents: true,
  displayedToStudentsAs: '',
  name: '',
  role: InstructorPermissionRole.INSTRUCTOR_PERMISSION_ROLE_COOWNER,
  joinState: JoinState.NOT_JOINED,

  permission: {
    privilege: {
      canModifyCourse: true,
      canModifySession: true,
      canModifyStudent: true,
      canModifyInstructor: true,
      canViewStudentInSections: true,
      canModifySessionCommentsInSections: true,
      canViewSessionInSections: true,
      canSubmitSessionInSections: true,
    },
    sectionLevel: [],
  },

  isEditing: true,
  isSavingInstructorEdit: false,
};

describe('InstructorCourseEditPageComponent', () => {
  let component: InstructorCourseEditPageComponent;
  let fixture: ComponentFixture<InstructorCourseEditPageComponent>;
  let courseService: CourseService;
  let instructorService: InstructorService;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [
        InstructorCourseEditPageComponent,
        InstructorEditPanelComponent,
        ViewRolePrivilegesModalComponent,
        CustomPrivilegeSettingPanelComponent,
        CopyInstructorsFromOtherCoursesModalComponent,
        CourseEditFormComponent,
      ],
      imports: [
        NgbModule,
        FormsModule,
        AjaxLoadingModule,
        TeammatesCommonModule,
        RouterTestingModule,
        HttpClientTestingModule,
        LoadingSpinnerModule,
        LoadingRetryModule,
        TeammatesRouterModule,
        SimpleModalModule,
        BrowserAnimationsModule,
        PanelChevronModule,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(InstructorCourseEditPageComponent);
    component = fixture.componentInstance;
    courseService = TestBed.inject(CourseService);
    instructorService = TestBed.inject(InstructorService);
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should load correct course details for given API output', () => {
    jest.spyOn(courseService, 'getCourseAsInstructor').mockReturnValue(of(testCourse));

    component.loadCourseInfo();

    expect(component.courseFormModel.course.courseId).toBe('exampleId');
    expect(component.courseFormModel.course.courseName).toBe('Example Course');
    expect(component.courseFormModel.course.timeZone).toBe('UTC (UTC)');
    expect(component.courseFormModel.course.creationTimestamp).toBe(0);
    expect(component.courseFormModel.course.deletionTimestamp).toBe(1000);
    expect(component.hasCourseLoadingFailed).toBeFalsy();
  });

  it('should not change course details if CANCEL is requested', () => {
    component.courseFormModel.course = testCourse;
    component.isCourseLoading = false;
    component.courseFormModel.originalCourse = { ...component.courseFormModel.course };
    fixture.detectChanges();

    component.courseFormModel.isEditing = true;
    component.courseFormModel.course.courseName = 'Example Course Changed';
    fixture.detectChanges();

    const button: any = fixture.debugElement.nativeElement.querySelector('#btn-cancel-course');
    button.click();

    expect(component.courseFormModel.isEditing).toBeFalsy();
    expect(component.courseFormModel.course.courseName).toBe('Example Course');
  });

  it('should update course details if SAVE is requested', () => {
    component.courseFormModel.course = testCourse;
    component.isCourseLoading = false;
    fixture.detectChanges();

    component.courseFormModel.isEditing = true;
    component.courseFormModel.course.courseName = 'Example Course Changed';
    fixture.detectChanges();

    jest.spyOn(courseService, 'updateCourse').mockReturnValue(of({
      courseId: 'exampleId',
      courseName: 'Example Course Changed',
      isCourseDeleted: false,
      timeZone: 'UTC (UTC)',
      institute: 'Test institute',
      creationTimestamp: 0,
      deletionTimestamp: 1000,
    }));

    const button: any = fixture.debugElement.nativeElement.querySelector('#btn-save-course');
    button.click();

    expect(component.courseFormModel.isEditing).toBeFalsy();
    expect(component.courseFormModel.course.courseName).toBe('Example Course Changed');
  });

  it('should load correct instructors details for given API output', () => {
    jest.spyOn(instructorService, 'loadInstructors').mockReturnValue(of({
      instructors: [testInstructor1, testInstructor2],
    }));

    component.loadCourseInstructors();

    expect(component.instructorDetailPanels[0].originalInstructor).toEqual(testInstructor1);
    expect(component.instructorDetailPanels[1].originalInstructor).toEqual(testInstructor2);
    expect(component.isInstructorsLoading).toBeFalsy();
  });

  it('should not add instructor if CANCEL is requested', () => {
    component.courseFormModel.course = testCourse;
    component.isCourseLoading = false;
    component.instructorDetailPanels = [
      {
        originalInstructor: { ...testInstructor1 },
        originalPanel: component.getInstructorEditPanelModel(testInstructor1),
        editPanel: component.getInstructorEditPanelModel(testInstructor1),
      },
      {
        originalInstructor: { ...testInstructor2 },
        originalPanel: component.getInstructorEditPanelModel(testInstructor2),
        editPanel: component.getInstructorEditPanelModel(testInstructor2),
      },
    ];
    component.isAddingNewInstructor = true;
    component.newInstructorPanel = component.getInstructorEditPanelModel(testInstructor3);
    component.newInstructorPanel.isEditing = true;
    fixture.detectChanges();

    const button: any = fixture.debugElement.nativeElement
        .querySelector(`#btn-cancel-instructor-${component.instructorDetailPanels.length + 1}`);
    button.click();

    expect(component.isAddingNewInstructor).toBeFalsy();
  });

  it('should add instructor details', () => {
    jest.spyOn(instructorService, 'createInstructor')
      .mockImplementation((params: { courseId: string, requestBody: InstructorCreateRequest }) => of({
        courseId: params.courseId,
        email: params.requestBody.email,
        joinState: JoinState.NOT_JOINED,
        name: params.requestBody.name,
      }));

    component.courseFormModel.course = testCourse;
    component.courseId = testCourse.courseId;
    component.isCourseLoading = false;
    component.instructorDetailPanels = [
      {
        originalInstructor: { ...testInstructor1 },
        originalPanel: component.getInstructorEditPanelModel(testInstructor1),
        editPanel: component.getInstructorEditPanelModel(testInstructor1),
      },
      {
        originalInstructor: { ...testInstructor2 },
        originalPanel: component.getInstructorEditPanelModel(testInstructor2),
        editPanel: component.getInstructorEditPanelModel(testInstructor2),
      },
    ];
    component.isAddingNewInstructor = true;
    component.newInstructorPanel = component.getInstructorEditPanelModel(testInstructor3);
    component.newInstructorPanel.isEditing = true;
    fixture.detectChanges();

    const button: any = fixture.debugElement.nativeElement
        .querySelector(`#btn-save-instructor-${component.instructorDetailPanels.length + 1}`);
    button.click();

    expect(component.isAddingNewInstructor).toBeFalsy();
    expect(component.isSavingNewInstructor).toBeFalsy();
    expect(component.instructorDetailPanels.length).toBe(3);
    expect(component.instructorDetailPanels[2].originalInstructor).toEqual(testInstructor3);
    expect(component.newInstructorPanel).toEqual(emptyInstructorPanel);
  });

  it('should re-order if instructor is deleted', () => {
    jest.spyOn(instructorService, 'deleteInstructor').mockReturnValue(of({}));

    component.courseFormModel.course = testCourse;
    component.isCourseLoading = false;
    component.instructorDetailPanels = [
      {
        originalInstructor: { ...testInstructor1 },
        originalPanel: component.getInstructorEditPanelModel(testInstructor1),
        editPanel: component.getInstructorEditPanelModel(testInstructor1),
      },
      {
        originalInstructor: { ...testInstructor2 },
        originalPanel: component.getInstructorEditPanelModel(testInstructor2),
        editPanel: component.getInstructorEditPanelModel(testInstructor2),
      },
    ];

    component.deleteInstructor(0);
    fixture.detectChanges();

    // using document instead of fixture as modal gets added into the dom outside the viewRef
    const button: any = document.getElementsByClassName('modal-btn-ok').item(0);
    button.click();
    fixture.detectChanges();

    expect(component.instructorDetailPanels.length).toBe(1);
    expect(component.instructorDetailPanels[0].originalInstructor).toEqual(testInstructor2);
  });

  it('should re-send reminder email for new instructors', () => {
    const mockReminderFunction: jest.MockedFunction<any> = jest.fn((_: string, email: string) => of({
      message: `An email has been sent to ${email}`,
    }));
    jest.spyOn(courseService, 'remindInstructorForJoin').mockImplementation(mockReminderFunction);

    component.courseFormModel.course = testCourse;
    component.isCourseLoading = false;
    component.instructorDetailPanels = [
      {
        originalInstructor: { ...testInstructor1 },
        originalPanel: component.getInstructorEditPanelModel(testInstructor1),
        editPanel: component.getInstructorEditPanelModel(testInstructor1),
      },
      {
        originalInstructor: { ...testInstructor2 },
        originalPanel: component.getInstructorEditPanelModel(testInstructor2),
        editPanel: component.getInstructorEditPanelModel(testInstructor2),
      },
    ];
    fixture.detectChanges();

    let button: any = fixture.debugElement.nativeElement
        .querySelector(`#btn-resend-invite-${component.instructorDetailPanels.length}`);
    button.click();

    // using document instead of fixture as modal gets added into the dom outside the viewRef
    button = document.getElementsByClassName('modal-btn-ok').item(0);
    button.click();

    expect(mockReminderFunction).toBeCalledWith(testCourse.courseId, testInstructor2.email);
  });

  it('should snap with default fields', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with course details', () => {
    component.courseFormModel.course = testCourse;

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should snap when editing course details', () => {
    component.courseFormModel.isEditing = true;

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should snap with some instructor details', () => {
    const instructor: Instructor = {
      name: 'Instructor A',
      email: 'instructora@example.com',
      courseId: component.courseId,
      joinState: JoinState.JOINED,
    };

    component.instructorDetailPanels = [
      {
        originalInstructor: instructor,
        originalPanel: component.getInstructorEditPanelModel(instructor),
        editPanel: component.getInstructorEditPanelModel(instructor),
      },
    ];

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });
});
