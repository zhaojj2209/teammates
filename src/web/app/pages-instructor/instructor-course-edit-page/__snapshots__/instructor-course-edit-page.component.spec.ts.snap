// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstructorCourseEditPageComponent should snap when editing course details 1`] = `
<tm-instructor-course-edit-page
  CourseEditFormMode={[Function Object]}
  CoursesSectionQuestions={[Function Object]}
  EditMode={[Function Object]}
  FormValidator={[Function Object]}
  Sections={[Function Object]}
  __ngContext__={[Function Number]}
  allSections={[Function Array]}
  allSessions={[Function Array]}
  authService={[Function AuthService]}
  courseFormModel={[Function Object]}
  courseId="undefined"
  courseService={[Function CourseService]}
  currInstructorCoursePrivilege={[Function Object]}
  currInstructorGoogleId=""
  feedbackSessionsService={[Function FeedbackSessionsService]}
  hasCourseLoadingFailed="false"
  hasInstructorsLoadingFailed="false"
  instructorDetailPanels={[Function Array]}
  instructorService={[Function InstructorService]}
  isAddingNewInstructor="false"
  isCopyingInstructor="false"
  isCourseLoading={[Function Boolean]}
  isInstructorsLoading="false"
  isSavingNewInstructor="false"
  navigationService={[Function NavigationService]}
  newInstructorPanel={[Function Object]}
  ngbModal={[Function NgbModal]}
  resetCourseFormEvent={[Function EventEmitter_]}
  route={[Function ActivatedRoute]}
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  studentService={[Function StudentService]}
>
  <tm-loading-retry>
    <tm-loading-spinner>
      <div
        class="loading-container"
      >
        <div
          class="row spinner-border text-primary"
          role="status"
        />
        <div
          class="row text-secondary"
        >
          Loading...
        </div>
      </div>
    </tm-loading-spinner>
  </tm-loading-retry><tm-loading-retry>
    <div
      class="right top-padded"
    >
      <a
        class="small"
        href="/web/instructor/help?questionId=course-instructor-access&section=courses"
        rel="noopener noreferrer"
        target="_blank"
        tmrouterlink="/web/instructor/help"
      >
        <i
          aria-hidden="true"
          class="fa fa-info-circle"
        />
         More about configuring access permissions 
      </a>
    </div>
    <div>
    </div>
  </tm-loading-retry><div
    class="card-body text-center"
  >
    <button
      class="btn btn-success mx-2 my-1"
      id="btn-add-instructor"
      type="button"
    >
       Add New Instructor 
    </button>
    <button
      class="btn btn-info mx-2 my-1"
      id="btn-copy-instructor"
      type="button"
    >
      Copy Instructors from Other Courses 
    </button>
  </div>
</tm-instructor-course-edit-page>
`;

exports[`InstructorCourseEditPageComponent should snap with course details 1`] = `
<tm-instructor-course-edit-page
  CourseEditFormMode={[Function Object]}
  CoursesSectionQuestions={[Function Object]}
  EditMode={[Function Object]}
  FormValidator={[Function Object]}
  Sections={[Function Object]}
  __ngContext__={[Function Number]}
  allSections={[Function Array]}
  allSessions={[Function Array]}
  authService={[Function AuthService]}
  courseFormModel={[Function Object]}
  courseId="undefined"
  courseService={[Function CourseService]}
  currInstructorCoursePrivilege={[Function Object]}
  currInstructorGoogleId=""
  feedbackSessionsService={[Function FeedbackSessionsService]}
  hasCourseLoadingFailed="false"
  hasInstructorsLoadingFailed="false"
  instructorDetailPanels={[Function Array]}
  instructorService={[Function InstructorService]}
  isAddingNewInstructor="false"
  isCopyingInstructor="false"
  isCourseLoading={[Function Boolean]}
  isInstructorsLoading="false"
  isSavingNewInstructor="false"
  navigationService={[Function NavigationService]}
  newInstructorPanel={[Function Object]}
  ngbModal={[Function NgbModal]}
  resetCourseFormEvent={[Function EventEmitter_]}
  route={[Function ActivatedRoute]}
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  studentService={[Function StudentService]}
>
  <tm-loading-retry>
    <tm-loading-spinner>
      <div
        class="loading-container"
      >
        <div
          class="row spinner-border text-primary"
          role="status"
        />
        <div
          class="row text-secondary"
        >
          Loading...
        </div>
      </div>
    </tm-loading-spinner>
  </tm-loading-retry><tm-loading-retry>
    <div
      class="right top-padded"
    >
      <a
        class="small"
        href="/web/instructor/help?questionId=course-instructor-access&section=courses"
        rel="noopener noreferrer"
        target="_blank"
        tmrouterlink="/web/instructor/help"
      >
        <i
          aria-hidden="true"
          class="fa fa-info-circle"
        />
         More about configuring access permissions 
      </a>
    </div>
    <div>
    </div>
  </tm-loading-retry><div
    class="card-body text-center"
  >
    <button
      class="btn btn-success mx-2 my-1"
      id="btn-add-instructor"
      type="button"
    >
       Add New Instructor 
    </button>
    <button
      class="btn btn-info mx-2 my-1"
      id="btn-copy-instructor"
      type="button"
    >
      Copy Instructors from Other Courses 
    </button>
  </div>
</tm-instructor-course-edit-page>
`;

exports[`InstructorCourseEditPageComponent should snap with default fields 1`] = `
<tm-instructor-course-edit-page
  CourseEditFormMode={[Function Object]}
  CoursesSectionQuestions={[Function Object]}
  EditMode={[Function Object]}
  FormValidator={[Function Object]}
  Sections={[Function Object]}
  __ngContext__={[Function Number]}
  allSections={[Function Array]}
  allSessions={[Function Array]}
  authService={[Function AuthService]}
  courseFormModel={[Function Object]}
  courseId="undefined"
  courseService={[Function CourseService]}
  currInstructorCoursePrivilege={[Function Object]}
  currInstructorGoogleId=""
  feedbackSessionsService={[Function FeedbackSessionsService]}
  hasCourseLoadingFailed="false"
  hasInstructorsLoadingFailed="false"
  instructorDetailPanels={[Function Array]}
  instructorService={[Function InstructorService]}
  isAddingNewInstructor="false"
  isCopyingInstructor="false"
  isCourseLoading={[Function Boolean]}
  isInstructorsLoading="false"
  isSavingNewInstructor="false"
  navigationService={[Function NavigationService]}
  newInstructorPanel={[Function Object]}
  ngbModal={[Function NgbModal]}
  resetCourseFormEvent={[Function EventEmitter_]}
  route={[Function ActivatedRoute]}
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  studentService={[Function StudentService]}
>
  <tm-loading-retry>
    <tm-loading-spinner>
      <div
        class="loading-container"
      >
        <div
          class="row spinner-border text-primary"
          role="status"
        />
        <div
          class="row text-secondary"
        >
          Loading...
        </div>
      </div>
    </tm-loading-spinner>
  </tm-loading-retry><tm-loading-retry>
    <div
      class="right top-padded"
    >
      <a
        class="small"
        href="/web/instructor/help?questionId=course-instructor-access&section=courses"
        rel="noopener noreferrer"
        target="_blank"
        tmrouterlink="/web/instructor/help"
      >
        <i
          aria-hidden="true"
          class="fa fa-info-circle"
        />
         More about configuring access permissions 
      </a>
    </div>
    <div>
    </div>
  </tm-loading-retry><div
    class="card-body text-center"
  >
    <button
      class="btn btn-success mx-2 my-1"
      id="btn-add-instructor"
      type="button"
    >
       Add New Instructor 
    </button>
    <button
      class="btn btn-info mx-2 my-1"
      id="btn-copy-instructor"
      type="button"
    >
      Copy Instructors from Other Courses 
    </button>
  </div>
</tm-instructor-course-edit-page>
`;

exports[`InstructorCourseEditPageComponent should snap with some instructor details 1`] = `
<tm-instructor-course-edit-page
  CourseEditFormMode={[Function Object]}
  CoursesSectionQuestions={[Function Object]}
  EditMode={[Function Object]}
  FormValidator={[Function Object]}
  Sections={[Function Object]}
  __ngContext__={[Function Number]}
  allSections={[Function Array]}
  allSessions={[Function Array]}
  authService={[Function AuthService]}
  courseFormModel={[Function Object]}
  courseId="undefined"
  courseService={[Function CourseService]}
  currInstructorCoursePrivilege={[Function Object]}
  currInstructorGoogleId=""
  feedbackSessionsService={[Function FeedbackSessionsService]}
  hasCourseLoadingFailed="false"
  hasInstructorsLoadingFailed="false"
  instructorDetailPanels={[Function Array]}
  instructorService={[Function InstructorService]}
  isAddingNewInstructor="false"
  isCopyingInstructor="false"
  isCourseLoading={[Function Boolean]}
  isInstructorsLoading="false"
  isSavingNewInstructor="false"
  navigationService={[Function NavigationService]}
  newInstructorPanel={[Function Object]}
  ngbModal={[Function NgbModal]}
  resetCourseFormEvent={[Function EventEmitter_]}
  route={[Function ActivatedRoute]}
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  studentService={[Function StudentService]}
>
  <tm-loading-retry>
    <tm-loading-spinner>
      <div
        class="loading-container"
      >
        <div
          class="row spinner-border text-primary"
          role="status"
        />
        <div
          class="row text-secondary"
        >
          Loading...
        </div>
      </div>
    </tm-loading-spinner>
  </tm-loading-retry><tm-loading-retry>
    <div
      class="right top-padded"
    >
      <a
        class="small"
        href="/web/instructor/help?questionId=course-instructor-access&section=courses"
        rel="noopener noreferrer"
        target="_blank"
        tmrouterlink="/web/instructor/help"
      >
        <i
          aria-hidden="true"
          class="fa fa-info-circle"
        />
         More about configuring access permissions 
      </a>
    </div>
    <div>
      <tm-instructor-edit-panel>
        <div
          class="card bg-light top-padded"
        >
          <div
            class="card-header bg-primary text-white text-center"
          >
            <strong>
              Instructor 1
            </strong>
            <div
              class="card-header-btn-toolbar"
            >
              <button
                class="btn btn-primary btn-sm"
                id="btn-edit-instructor-1"
              >
                <i
                  aria-hidden="true"
                  class="fa fa-pen"
                />
                 Edit 
              </button>
              <button
                class="btn btn-primary btn-sm"
                id="btn-delete-instructor-1"
                ngbtooltip="Delete the instructor from the course"
                placement="top-right"
              >
                <i
                  aria-hidden="true"
                  class="fa fa-trash"
                />
                 Delete 
              </button>
            </div>
          </div>
          <div
            class="card-body text-center"
          >
            <div
              class="row form-group"
            >
              <label
                class="col-sm-2 control-label"
                for="google-id-instructor-1"
              >
                 Google ID: 
              </label>
              <div
                class="col-sm-9"
              >
                <input
                  class="form-control ng-untouched ng-pristine ng-valid"
                  disabled=""
                  id="google-id-instructor-1"
                  maxlength="254"
                />
              </div>
            </div>
            <div
              class="row form-group"
            >
              <label
                class="col-sm-2 control-label"
                for="name-instructor-1"
              >
                 Name: 
              </label>
              <div
                class="col-sm-9"
              >
                <input
                  class="form-control ng-untouched ng-pristine ng-valid"
                  id="name-instructor-1"
                  maxlength="100"
                />
              </div>
            </div>
            <div
              class="row form-group"
            >
              <label
                class="col-sm-2 control-label"
                for="email-instructor-1"
              >
                 Email: 
              </label>
              <div
                class="col-sm-9"
              >
                <input
                  class="form-control ng-untouched ng-pristine ng-valid"
                  id="email-instructor-1"
                  maxlength="254"
                />
              </div>
            </div>
            <div
              class="row form-group"
            >
              <label
                class="col-sm-2 control-label"
                for="displayed-name-instructor-1"
              >
                <input
                  aria-label="Enable display name"
                  checked=""
                  class="ng-untouched ng-pristine ng-valid"
                  id="checkbox-display-instructor-1"
                  type="checkbox"
                />
                <span
                  class="ngb-tooltip-class"
                  ngbtooltip="If this is unselected, the instructor will be completely invisible to students. E.g. to give access to a colleague for ‘auditing’ your course"
                >
                  Display to students as:
                </span>
              </label>
              <div
                class="col-sm-9"
              >
                <input
                  class="form-control ng-untouched ng-pristine ng-valid"
                  id="displayed-name-instructor-1"
                  placeholder="E.g. Co-lecturer, Teaching Assistant (default: Instructor)"
                  type="text"
                  value="Instructor"
                />
              </div>
            </div>
            <div
              class="row form-group align-items-start text-start"
            >
              <label
                class="col-sm-2 control-label text-center text-md-end"
              >
                 Access Level: 
              </label>
              <div
                class="col-sm-9"
                id="access-levels-instructor-1"
              >
                <span
                  id="role-instructor-1"
                >
                  Custom: No access by default. Any access needs to be granted explicitly.
                </span>
                <button
                  class="btn btn-link padding-0"
                  type="button"
                >
                  View Details
                </button>
              </div>
            </div>
          </div>
        </div>
      </tm-instructor-edit-panel>
      <br />
    </div>
  </tm-loading-retry><div
    class="card-body text-center"
  >
    <button
      class="btn btn-success mx-2 my-1"
      id="btn-add-instructor"
      type="button"
    >
       Add New Instructor 
    </button>
    <button
      class="btn btn-info mx-2 my-1"
      id="btn-copy-instructor"
      type="button"
    >
      Copy Instructors from Other Courses 
    </button>
  </div>
</tm-instructor-course-edit-page>
`;
